### ISSUE TITLE
Util constant analysis should not depend on LinalgExt dialect

### ISSUE THREAD

**Issue by hanhanW** [iree-org/iree#14887]
We need a trait/interface to describe what should be hoisted and not be hoisted. Then we don't need a dep between Util and LinalgExt dialect.

**Comment by hanhanW** [comment]
I'm considering if we should just move those ops from LinalgExt to Util dialect. In this context, constant analysis no longer depend on LinalgExt dialect.

https://github.com/openxla/iree/blob/4bb3848e38f5249479f367c3f95054b857528ee6/compiler/src/iree/compiler/Dialect/Util/Analysis/Constant/OpOracle.cpp#L104-L109

**Comment by benvanik** [comment]
we don't want these in util - util is like STL/libc - if anything, they should exist in codegen (or a codegen-adjacent dialect)
to decouple (regardless of where they go) we need the trait/interface

**Comment by benvanik** [comment]
are these ops strictly used in flow/? if so they could go there, but I'd want to rename them before doing it (and ensure we want them as part of the permanent flow opset)

**Comment by hanhanW** [comment]
I thought we want `set_encoding` and `unset_encoding` in utils. Today they are only used by GlobalOptimization and CodeGen. If we are going to make `set_encoding` carry padding semantics, flow/stream need to learn about encoding types.

**Comment by hanhanW** [comment]
The `UpperBoundTileSizeOp` will be removed if we make encoding types have padding semantics.

**Comment by benvanik** [comment]
as mentioned util is like the stl - it's used in every dialect (input, flow, stream, hal, custom dialects, etc) - stuff that only exists in specific layers doesn't belong there. we may need a dumping ground for ops that live at that layer - tensor_ext (which is what linalg_ext is :) - util ain't it though :)

**Comment by benvanik** [comment]
(btw plz loop me into any discussions related to stream - flow is pretty flexible but padding semantics making it to stream are tricky and I want to make sure we don't end up with something we can't lower - we have a nice place to do it but physical storage encodings are likely going to be needed to represent such things and that'll be a bit bigger in scope - best if we can bake it out beforehand)

**Comment by hanhanW** [comment]
Yes, we will definitely loop you in beforehand when we start working on it! I heard the idea before, but I don't know the details. I'm just thinking about possible solutions when revisiting old issues. :-D

**Comment by hanhanW** [comment]
It's fixed by https://github.com/iree-org/iree/commit/8c2a3a9cb9f9bb97bc9c7d03be88bfd87e127940
